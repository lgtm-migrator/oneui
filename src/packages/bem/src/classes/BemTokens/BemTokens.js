import BemPrefixSet from '../BemPrefixSet';

/**
 * @typedef {Object} BemEntity
 * @property {string} block - Block name
 * @property {string|undefined} elem - Elem name
 * @property {string|undefined} mod - Mod name
 * @property {string|undefined} value - Mod value
 */

export class BemTokensError extends Error {}

export default class BemTokens {
    constructor(tokens) {
        const { block, elem, mod, value } = tokens;
        Object.assign(this, { block, elem, mod, value });
    }

    /**
     * @returns {RegExp} - Pattern for matching and parsing a BEM classname
     */
    static createClassNamePattern(prefixes) {
        const bemEntityPattern = '[a-zA-Z]+[a-zA-Z0-9]*'; // Pattern for matching block name, elem name or modifier name
        const valuePattern = '[a-zA-Z0-9]+'; // Pattern for matching modifier's value
        const { elem, mod, value } = prefixes;
        const pattern = new RegExp(
            `^(${bemEntityPattern})` +
                `(${elem}${bemEntityPattern})?` +
                `(${mod}${bemEntityPattern})?` +
                `(${value}${valuePattern})?$`
        );
        return pattern;
    }

    /**
     * @param {string} value - Value to be stripped from a prefix
     * @param {string} prefix - Prefix
     * @returns {string} - Stripped value
     */
    static stripPrefix(value, prefix) {
        if (typeof value !== 'string') return '';
        if (value.startsWith(prefix)) {
            return value.replace(prefix, '');
        }
        return value;
    }

    static validateClassName(className) {
        if (typeof className !== 'string') {
            throw new BemTokensError('BEM classname shoud be a string');
        }
        if (className === '') {
            throw new BemTokensError('BEM classname shoud not be empty');
        }
    }

    static from(className = '', prefixes = {}) {
        BemTokens.validateClassName(className);
        const prefixSet = new BemPrefixSet(prefixes);
        const classNamePattern = BemTokens.createClassNamePattern(prefixSet);
        const match = className.match(classNamePattern);
        if (match === null) {
            throw new BemTokensError(`BEM classname '${className}' has invalid syntax or empty`);
        }
        const [, block, elem, mod, value] = match;
        return new BemTokens({
            block,
            elem: BemTokens.stripPrefix(elem, prefixSet.elem),
            mod: BemTokens.stripPrefix(mod, prefixSet.mod),
            value: BemTokens.stripPrefix(value, prefixSet.value)
        });
    }
}
