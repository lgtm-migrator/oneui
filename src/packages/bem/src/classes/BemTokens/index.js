export { default, BemTokensError } from './BemTokens';
