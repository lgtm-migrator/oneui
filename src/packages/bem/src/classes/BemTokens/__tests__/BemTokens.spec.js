import BemTokens, { BemTokensError } from '..';

describe('BemTokens', () => {
    describe('module export', () => {
        it('should export BemTokens as default', () => {
            expect(BemTokens).toBeTruthy();
        });
        it('should export BemTokensError as named export', () => {
            expect(BemTokensError).toBeTruthy();
        });
    });

    describe('.from', () => {
        it('should parse "block" correctly', () => {
            const bemTokens = BemTokens.from('block');
            expect(bemTokens).toBeInstanceOf(BemTokens);
            expect(bemTokens).toEqual({
                block: 'block',
                elem: '',
                mod: '',
                value: ''
            });
        });
        it('should parse "block--mod" correctly', () => {
            const bemTokens = BemTokens.from('block--mod');
            expect(bemTokens).toEqual({
                block: 'block',
                elem: '',
                mod: 'mod',
                value: ''
            });
        });
        it('should parse "block--mod_value" correctly', () => {
            const bemTokens = BemTokens.from('block--mod_value');
            expect(bemTokens).toEqual({
                block: 'block',
                elem: '',
                mod: 'mod',
                value: 'value'
            });
        });
        it('should parse "block__elem" correctly', () => {
            const bemTokens = BemTokens.from('block__elem');
            expect(bemTokens).toEqual({
                block: 'block',
                elem: 'elem',
                mod: '',
                value: ''
            });
        });
        it('should parse "block__elem--mod" correctly', () => {
            const bemTokens = BemTokens.from('block__elem--mod');
            expect(bemTokens).toEqual({
                block: 'block',
                elem: 'elem',
                mod: 'mod',
                value: ''
            });
        });
        it('should parse "block__elem--mod_value" correctly', () => {
            const bemTokens = BemTokens.from('block__elem--mod_value');
            expect(bemTokens).toEqual({
                block: 'block',
                elem: 'elem',
                mod: 'mod',
                value: 'value'
            });
        });
        it('should throw if className is empty or not a string', () => {
            expect(() => BemTokens.from()).toThrowErrorMatchingSnapshot();
            expect(() => BemTokens.from(42)).toThrowErrorMatchingSnapshot();
            expect(() => BemTokens.from({ foo: 'bar ' })).toThrowErrorMatchingSnapshot();
        });
    });
});
