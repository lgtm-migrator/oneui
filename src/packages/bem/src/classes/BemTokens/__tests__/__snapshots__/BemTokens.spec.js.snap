// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BemTokens .from should throw if className is empty or not a string 1`] = `"BEM classname shoud not be empty"`;

exports[`BemTokens .from should throw if className is empty or not a string 2`] = `"BEM classname shoud be a string"`;

exports[`BemTokens .from should throw if className is empty or not a string 3`] = `"BEM classname shoud be a string"`;

exports[`BemTokens .from should throw if className is empty or not of a string type 1`] = `"BEM classname shoud not be empty"`;

exports[`BemTokens .from should throw if className is empty or not of a string type 2`] = `"BEM classname shoud be a string"`;

exports[`BemTokens .from should throw if className is empty or not of a string type 3`] = `"BEM classname shoud be a string"`;
